lib/sycamore/sycamore/transforms/similarity.py

```diff
@@ -0,0 +1,224 @@
+
+        return doc_batch
+
+    @abstractmethod
+    def score(self, inputs: list[tuple[str, str]]) -> list[float]:
+        pass
+
+
+class HuggingFaceTransformersSimilarityScorer(SimilarityScorer):
+    """
+    HuggingFaceTransformersSimilarityScorer is an SimilarityScorer class for generating sentence similarity using the
+    transformers package.
+
+    Args:
+        model_name: Name or path of the Transformers model to use for similarity scoring.
+        model_batch_size: Batch size used by the underlying Transformers model for similarity scoring, default is 16.
+        max_tokens: Max tokens to use for tokenization, default is 512.
+        device: Device (e.g., "cpu" or "cuda") on which to perform embedding.
+        ignore_doc_structure: Ignore Document model (Document->Elements) in a DocSet
+
+    Example:
+        .. code-block:: python
+
+            similarity_scorer = HuggingFaceTransformersSimilarityScorer(ignore_doc_structure=True)
+
+            doc = Document(
+                {
+                    "doc_id": 1,
+                    "elements": [
+                        {"text_representation": "here is an animal with 4 legs and whiskers"},
+                    ],
+                }
+            )
+            result = similarity_scorer.generate_similarity_scores(
+                [doc], query="is this a cat?", score_property_name="similarity_score"
+            )
+            print(result.properties["similarity_score"])
+
+    """
+
+    def __init__(
+        self,
+        model_name: str = "BAAI/bge-reranker-large",
+        model_batch_size: int = 16,
+        max_tokens: int = 512,
+        device: Optional[str] = None,
+        ignore_doc_structure: bool = False,
+        ignore_element_sources: Optional[list[DocumentSource]] = None,
+    ):
+        super().__init__(ignore_element_sources=ignore_element_sources, ignore_doc_structure=ignore_doc_structure)
+        self.device = choose_device(device)
+        self.model_name = model_name
+        self.model_batch_size = model_batch_size
+        self.max_tokens = max_tokens
+
+        self._model = None
+        self._tokenizer = None
+
+    @requires_modules(["transformers", "torch"], extra="local-inference")
+    def __call__(self, doc_batch: list[Document], query: str, score_property_name: str) -> list[Document]:
+        return self.generate_similarity_scores(doc_batch, query, score_property_name)
+
+    @timetrace("TransformersSimilarity")
+    def score(self, inputs: list[tuple[str, str]]) -> list[float]:
+        import torch
+
+        if not self._model or not self._tokenizer:
+            logger.info(f"Initializing transformers model: {self.model_name}")
+            from transformers import AutoModelForSequenceClassification, AutoTokenizer
+
+            self._tokenizer = AutoTokenizer.from_pretrained(self.model_name)
+            self._model = AutoModelForSequenceClassification.from_pretrained(self.model_name).to(self.device)
```

Comment:
Do you need to do any explicit stateful actor things bc of this? I guess since it's a class you're good. idk I kinda lost track of how you differentiate them in all of eric's ray-removal

Reply:
No clue, @eric-anderson anything to note here? We're initializing the model once per class instance